### PR TITLE
lib: nrf_modem: net_if: Kconfig changes and IPv4 only support

### PR DESCRIPTION
**Pull request by eivindj-nordic**
* Replace build asserts with Kconfig dependencies.
* Allow supporting IPv4 only, without IPv6. It is up to the application to make sure this is acceptable to their use. IPv6 is enabled by default.

### CODE REVIEW

File: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+.. _nrf_modem_lib_lte_net_if:
+
+LTE Network Interface
```
Comment (lemrey) [inline doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:3]: `LTE network interface` is the name of the Zephyr network interface driver for the nrf91 Series modem firmware.

I would call the chapter `Zephyr network interface driver` or `Network interface driver`, rather than `LTE network interface`, and begin explaining with:

```
The LTE network interface in the Modem library is a Zephyr network interface driver for the nRF91 Series cellular firmware.
By enabling the LTE network interface in the Modem library, you can control the nRF91 series modem as a Zephyr network interface, for example, :c:func:`net_if_up` and :c:func`net_if_down`.
```

File: lib/nrf_modem_lib/lte_net_if/lte_ip_addr_helper.c [hunk 2]
```diff
@@ -49,15 +53,34 @@ static int ip_addr_get(char *addr4, char *addr6)
...
+	if ((!addr4 && !addr6) || (addr4 && addr6)) {
+		return -EINVAL;
+	}
+
+#if !CONFIG_NET_IPV6
+	if (!addr4) {
+		LOG_ERR("IPv6 address requested when IPv6 is not enabled");
+		return -EINVAL;
+	}
+#endif /* !CONFIG_NET_IPV6 */
```
Comment (lemrey) [inline lib/nrf_modem_lib/lte_net_if/lte_ip_addr_helper.c:69]: This is okay, but would it not be a bit more straightforward to check
```
#if !CONFIG_NET_IPV6
if (addr6) {
    LOG_ERR("IPv6 address requested when IPv6 is not enabled");
}
#endif
```

File: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+.. contents::
+   :local:
+   :depth: 2
+
+The LTE Network interface is intended to be integrated with and bound to a Zephyr network interface via the Connection Manager binding structure :ref:`conn_mgr_conn_api`.
+It can be managed via Zephyr's network management API and the Connection Manager Connectivity API using calls such as the :c:func:`net_if_down`, :c:func:`net_if_up`, and :c:func:`conn_mgr_if_connect` functions.
+
+.. note::
+
+   The nRF modem supports both IPv6 and IPv4. At any time, either IP families can be given by the network, independently of each other.
```
Comment (lemrey) [inline doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:15]: We should not call it the `nRF modem`

Comment (lemrey) [reply doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:15]: I would rephrase this entire part by saying just

```
The LTE network interface supports both IPv4 and IPv6.
The support is enabled automatically for whichever IP stack version is enabled in Zephyr.
```

Comment (lemrey) [reply doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:15]: And this brings me to a question.. should we allow to disable IPv4 and use only IPv6?

Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst]: It's not a big change so I can add that.

File: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+Network interface driver
+########################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The LTE network interface in the Modem library is a Zephyr network interface driver for the nRF91 Series cellular firmware.
+By enabling the LTE network interface in the Modem library, you can control the nRF91 series modem as a Zephyr network interface, for example, :c:func:`net_if_up`, :c:func:`conn_mgr_if_connect` and :c:func:`net_if_down`.
```
Comment (MirkoCovizzi) [inline doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:11]:
```suggestion
By enabling the LTE network interface in the Modem library, you can control the nRF91 series modem as a Zephyr network interface, for example, :c:func:`net_if_up`, :c:func:`conn_mgr_if_connect`, and :c:func:`net_if_down`.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 5]
```diff
@@ -796,6 +796,12 @@ Modem libraries
   * Fixed an issue with the CFUN hooks when the Modem library is initialized during ``SYS_INIT`` at kernel level and makes calls to the :ref:`nrf_modem_at` interface before the application level initialization is done.
   * Removed the deprecated options ``CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_ASYNC`` and ``CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_SYNC``.
 
+    * ``lte_net_if`` network interface driver:
+
+    * Added the dependency to the :kconfig:option:`CONFIG_NET_CONNECTION_MANAGER` Kconfig option.
```
Comment (MirkoCovizzi) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:801]: I think that this section should be indented.

File: lib/nrf_modem_lib/lte_net_if/lte_ip_addr_helper.c [hunk 6]
```diff
@@ -49,25 +53,54 @@ static int ip_addr_get(char *addr4, char *addr6)
...
+	if ((!addr4 && !addr6) || (addr4 && addr6)) {
+		return -EINVAL;
+	}
+
+#if !CONFIG_NET_IPV4
+	if (addr4) {
+		LOG_ERR("IPv4 address requested when IPv4 is not enabled");
+		return -EINVAL;
+	}
+#endif /* !CONFIG_NET_IPV6 */
```
Comment (MirkoCovizzi) [inline lib/nrf_modem_lib/lte_net_if/lte_ip_addr_helper.c:69]:
```suggestion
#endif /* !CONFIG_NET_IPV4 */
```

**Review by divipillai — APPROVED** [review]

File: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst [hunk 7]
```diff
@@ -0,0 +1,13 @@
+.. _nrf_modem_lib_lte_net_if:
+
+Network interface driver
+########################
+
+The LTE network interface in the  :ref:`nrfxlib:nrf_modem` is a Zephyr network interface driver for the nRF91 Series cellular firmware.
+
+By enabling the LTE network interface in the Modem library, you can control the nRF91 series modem as a Zephyr network interface, for example, :c:func:`net_if_up`, :c:func:`conn_mgr_if_connect`, and :c:func:`net_if_down`.
```
Comment (MirkoCovizzi) [inline doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_lte_net_if.rst:8]: I suppose it should be like this here too. Also added a missing verb (adjust to your taste).
```suggestion
By enabling the LTE network interface in the :ref:`nrfxlib:nrf_modem`, you can control the nRF91 series modem as a Zephyr network interface, for example by calling :c:func:`net_if_up`, :c:func:`conn_mgr_if_connect`, and :c:func:`net_if_down`.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 8]
```diff
@@ -808,12 +808,23 @@ Modem libraries
 
 * :ref:`nrf_modem_lib_readme`:
 
-  * Added the Kconfig option  :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_CHUNK_SZ` to process traces in chunks.
-    This can improve the availability of trace memory, and thus reduce the chances of losing traces.
+  * Added:
+
+    * The Kconfig option  :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_CHUNK_SZ` to process traces in chunks.
```
Comment (MirkoCovizzi) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:813]:
```suggestion
    * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_UART_CHUNK_SZ` Kconfig option to process traces in chunks.
```

**Review by MirkoCovizzi — APPROVED** [review]

File: doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_fault.rst [hunk 9]
```diff
@@ -16,6 +16,7 @@ The behavior of the :c:func:`nrf_modem_fault_handler` function is controlled wit
 * :kconfig:option:`CONFIG_NRF_MODEM_LIB_ON_FAULT_DO_NOTHING` - This option lets the fault handler log the Modem fault and return (default).
 * :kconfig:option:`CONFIG_NRF_MODEM_LIB_ON_FAULT_RESET_MODEM` - This option lets the fault handler schedule a workqueue task to reinitialize the modem and Modem library.
 * :kconfig:option:`CONFIG_NRF_MODEM_LIB_ON_FAULT_APPLICATION_SPECIFIC` - This option lets the fault handler function :c:func:`nrf_modem_fault_handler` be defined by the application, outside of the Modem library integration layer.
+* :kconfig:option:`CONFIG_NRF_MODEM_LIB_ON_FAULT_LTE_NET_IF` - This option lets the :ref:`nrf_modem_lib_lte_net_if` handle the fault. The application can be notified about the fatal error by subscribing to the ``NET_EVENT_CONN_IF_FATAL_ERROR`` event using the :c:func:`net_mgmt_init_event_callback` function.
```
Comment (lemrey) [inline doc/nrf/libraries/modem/nrf_modem_lib/nrf_modem_lib_fault.rst:19]: Rather, this option is the only option available for the user to handle the fault when using `lte_net_if`. I think this should be a bit more explicit. What it does is to signal the fault with `NET_EVENT_CONN_IF_FATAL_ERROR`, which the application subscribes to using `net_mgmt_init_event_callback`.

Perhaps:
```
This is the only option available when CONFIG_NRF_MODEM_LIB_NET_IF is selected.
With this option, modem faults are signaled to the application as fatal network events (NET_EVENT_CONN_IF_FATAL_ERROR), to which the application subscribes by calling the 
net_mgmt_init_event_callback() function.
```

**Review by lemrey — COMMENTED** [review]
Looks good, left a comment about the documentation.
Lots of stuff happening in one commit, if possible consider splitting it.

**Review by divipillai — APPROVED** [review]
